customcommands/tmplextensions.go

```diff
@@ -488,6 +497,97 @@ func tmplDBDelById(ctx *templates.Context) interface{} {
 		return "", err
 	}
 }
+func tmplDBDelMultiple(ctx *templates.Context) interface{} {
+	return func(query interface{}, iAmount interface{}, iSkip interface{}) (interface{}, error) {
+		if ctx.IncreaseCheckCallCounterPremium("db_interactions", 10, 50) {
+			return "", templates.ErrTooManyCalls
+		}
+
+		if ctx.IncreaseCheckCallCounterPremium("db_multiple", 2, 10) {
+			return "", templates.ErrTooManyCalls
+		}
+		
+		q, err := queryFromArg(query)
+		if err!= nil {
+			return "", err
+		}
+		
+		amount := int(templates.ToInt64(iAmount))
+		if amount > 100 {
+			amount = 100
+		}
+		skip := int(templates.ToInt64(iSkip))
+		orderby := "value_num DESC, id DESC"
+		if q.Reverse {
+			orderby = "value_num ASC, id ASC"
+		}
+
+		qms := []qm.QueryMod{qm.Where("guild_id = ?", ctx.GS.ID), qm.OrderBy(orderby), qm.Limit(amount), qm.Offset(skip)}
+		if q.Pattern.Valid {
+			qms = append(qms, qm.Where("key LIKE ?", limitString(q.Pattern.String, 256)))
+		}
+		if q.UserID.Valid {
+			qms = append(qms, qm.Where("user_id = ?", q.UserID.Int64))
+		}
+		rows, err := models.TemplatesUserDatabases(qms...).AllG(context.Background())
+		if err != nil {
+			return "", err
+		}
+		
+		cleared, err := rows.DeleteAllG(context.Background())
+		ctx.GS.UserCacheDel(CacheKeyDBLimits)
+		return cleared, err
+	}
+}
+
+func tmplDBRank(ctx *templates.Context) interface{} {
+	return func(query interface{}, userID int64, key string) (interface{}, error) {
+		if ctx.IncreaseCheckCallCounterPremium("db_interactions", 10, 50) {
+			return "", templates.ErrTooManyCalls
+		}
+
+		if ctx.IncreaseCheckCallCounterPremium("db_multiple", 2, 10) {
+			return "", templates.ErrTooManyCalls
+		}
+		
+		q, err := queryFromArg(query)
+		if err!= nil {
+			return "", err
+		}	
+		
+		order := `DESC`
+		if q.Reverse {
+			order = `ASC`
+		}
+		
+		if q.UserID.Valid && (q.UserID.Int64 != userID) { // some optimization
+			return 0, nil
+		}
+
+		const rawquery = `SELECT position FROM
+(
+	SELECT user_id, key,
+	RANK() OVER
+	(
+		ORDER BY 
+		CASE WHEN $1 = 'ASC'  THEN value_num ELSE 0 END ASC,  
+		CASE WHEN $1 = 'DESC' THEN value_num ELSE 0 END DESC,
+		CASE WHEN $1 = 'ASC'  THEN id ELSE 0 END ASC,
+		CASE WHEN $1 = 'DESC'  THEN id ELSE 0 END DESC
+	) AS position
+FROM templates_user_database WHERE (guild_id = $2) AND ($3::bigint IS NULL OR user_id = $3) AND ($4::text IS NULL OR key LIKE $4) AND (expires_at IS NULL OR expires_at > now())
+) AS w
+WHERE user_id = $5 AND key = $6`
+
+		var rank int64
+		err = common.PQ.QueryRow(rawquery, order, ctx.GS.ID, q.UserID, q.Pattern, userID, key).Scan(&rank)
```

Comment:
why is there 2 userID's?

Reply:
so this function is used to find the rank of a particular entry (defined by userID and key) over a particular set of entries or the entire database. When no pattern/userID is mentioned, it finds the global rank. The pattern/userID can be used to further narrow down the set of entries on which the rank is to be found. Now this can be handled using an userID bool in this particular scenario, however for consistency accross functions I have reused the same query structure as being used in the other two functions. Indeed this means if the 2 userIDs here dont match, there is no point in performing a query and I have already handled this special case for optimization.